skfp/distances/rand.py

```diff
@@ -166,3 +167,167 @@ def rand_binary_distance(
     0.33333333333333337
     """
     return 1 - rand_binary_similarity(vec_a, vec_b)
+
+
+@validate_params(
+    {"X": ["array-like"], "Y": ["array-like", None]},
+    prefer_skip_nested_validation=True,
+)
+def bulk_rand_binary_similarity(
+    X: np.ndarray, Y: Optional[np.ndarray] = None
+) -> np.ndarray:
+    r"""
+    Bulk Rand similarity for binary matrices.
+
+    Computes the pairwise Rand [1]_ [2]_ (known as All-Bit [3]_ or Sokal-Michener)
+    similarity between binary matrices. If one array is passed, similarities are
+    computed between its rows. For two arrays, similarities are between their respective
+    rows, with `i`-th row and `j`-th column in output corresponding to `i`-th row from
+    first array and `j`-th row from second array.
+
+    See also :py:func:`rand_binary_similarity`.
+
+    Parameters
+    ----------
+    X : ndarray
+        First binary input array, of shape :math:`m \times m`
+
+    Y : ndarray, default=None
+        Second binary input array, of shape :math:`n \times n`. If not passed, similarities
+        are computed between rows of X.
+
+    Returns
+    -------
+    similarities : ndarray
+        Array with pairwise Rand similarity values. Shape is :math:`m \times n` if two
+        arrays are passed, or :math:`m \times m` otherwise.
+
+    References
+    ----------
+    .. [1] `Rand, W.M.
+        "Objective criteria for the evaluation of clustering methods."
+        J. Amer. Stat. Assoc. 1971; 66: 846–850.
+        <https://www.tandfonline.com/doi/abs/10.1080/01621459.1971.10482356>`_
+
+    .. [2] `Deza M.M., Deza E.
+        "Encyclopedia of Distances."
+        Springer, Berlin, Heidelberg, 2009.
+        <https://doi.org/10.1007/978-3-642-00234-2_1>`_
+
+    .. [3] `RDKit documentation
+        <https://www.rdkit.org/docs/source/rdkit.DataStructs.cDataStructs.html>`_
+
+    See Also
+    --------
+    :py:func:`rand_binary_similarity` : Rand similarity function for two vectors.
+
+    Examples
+    --------
+    >>> from skfp.distances import bulk_rand_binary_similarity
+    >>> import numpy as np
+    >>> X = np.array([[1, 0, 1], [0, 0, 1]])
+    >>> Y = np.array([[1, 0, 1], [0, 1, 1]])
+    >>> sim = bulk_rand_binary_similarity(X, Y)
+    >>> sim
+    array([[0.66666667, 0.33333333],
+           [0.33333333, 0.33333333]])
+    """
+    if Y is None:
+        return _bulk_rand_binary_similarity_single(X)
+    else:
+        return _bulk_rand_binary_similarity_two(X, Y)
+
+
+@numba.njit(parallel=True)
+def _bulk_rand_binary_similarity_single(X: np.ndarray) -> np.ndarray:
+    m = X.shape[0]
+    sims = np.empty((m, m))
+
+    # upper triangle - actual similarities
+    for i in numba.prange(m):
+        # in this case diagonal is not always 1
+        for j in numba.prange(i, m):
+            intersection = np.sum(np.logical_and(X[i], X[j]))
+            length = len(X[i])
+            sims[i, j] = intersection / length if length != 0 else 1.0
```

Comment:
Don't check for zero length. Such data is nonsensical, and it's user's responsibility then